python/src/iceberg/expression/literals.py

```diff
@@ -200,78 +205,52 @@ def _(self, type_var):
         return self
 
 
-class IntegerLiteral(Literal[int]):
-    @singledispatchmethod
-    def to(self, type_var):
-        return None
-
-    @to.register(IntegerType)
-    def _(self, type_var: IntegerType) -> "IntegerLiteral":
-        return self
-
-    @to.register(LongType)
-    def _(self, type_var: LongType) -> "LongLiteral":
-        return LongLiteral(self.value)
-
-    @to.register(FloatType)
-    def _(self, type_var: FloatType) -> "FloatLiteral":
-        return FloatLiteral(float(self.value))
-
-    @to.register(DoubleType)
-    def _(self, type_var: DoubleType) -> "DoubleLiteral":
-        return DoubleLiteral(self.value)
-
-    @to.register(DateType)
-    def _(self, type_var: DateType) -> "DateLiteral":
-        return DateLiteral(self.value)
-
-    @to.register(DecimalType)
-    def _(self, type_var: DecimalType) -> "DecimalLiteral":
-        if type_var.scale == 0:
-            return DecimalLiteral(Decimal(self.value))
-        else:
-            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
-
-
 class LongLiteral(Literal[int]):
     @singledispatchmethod
     def to(self, type_var):
         return None
 
     @to.register(LongType)
-    def _(self, type_var: LongType) -> "LongLiteral":
+    def _(self, type_var: LongType) -> Literal[int]:
         return self
 
     @to.register(IntegerType)
-    def _(self, type_var: IntegerType) -> Union[AboveMax, BelowMin, IntegerLiteral]:
+    def _(self, type_var: IntegerType) -> Union[AboveMax, BelowMin, Literal[int]]:
         if IntegerType.max < self.value:
             return AboveMax()
         elif IntegerType.min > self.value:
             return BelowMin()
-        return IntegerLiteral(self.value)
+        return self
 
     @to.register(FloatType)
-    def _(self, type_var: FloatType) -> "FloatLiteral":
+    def _(self, type_var: FloatType) -> Literal[float]:
         return FloatLiteral(float(self.value))
 
     @to.register(DoubleType)
-    def _(self, type_var: DoubleType) -> "DoubleLiteral":
-        return DoubleLiteral(self.value)
+    def _(self, type_var: DoubleType) -> Literal[float]:
+        return DoubleLiteral(float(self.value))
```

Comment:
oh gotcha, this is about what you suggested [earlier on what's missing](https://github.com/apache/iceberg/pull/4262#discussion_r837642495), right? Thank you!

Reply:
Yes, that's what I was referring to. Looks like Python allows float and int comparison as long as the value has an exact representation in floating point. That's why the tests still works, but this change is still needed to make sure the inner type is correct.